File: SmaRTC-core/sdk/csharp/SmaRTCSimple.cs
// 💜 SmaRTC C# SDK Simplifié
// Wrapper moderne pour .NET 6+ et Unity
using System;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;
using System.Collections.Generic;

namespace SmaRTC.Client
{
    /// <summary>
    /// Client SmaRTC simplifié pour .NET
    /// 
    /// Exemple d'utilisation :
    /// <code>
    /// var client = new SmaRTCSimple();
    /// await client.LoginAsync("alice", "password123");
    /// var session = await client.StartCallAsync("Ma réunion");
    /// </code>
    /// </summary>
    public class SmaRTCSimple : IDisposable
    {
        // ====================================================================
        // Configuration
        // ====================================================================

        public class Config
        {
            public string ApiBaseUrl { get; set; } = "http://localhost:8080";
            public string SignalServerUrl { get; set; } = "http://localhost:5001";
            public int TimeoutSeconds { get; set; } = 10;
            public bool EnableLogs { get; set; } = false;
        }

        // ====================================================================
        // Types de données
        // ====================================================================

        public record User(string Id, string Username);

        public record Session(
            string SessionId,
            string RoomName,
            string HostUserId,
            List<string> Participants,
            DateTime CreatedAt,
            bool IsActive
        );

        public record ICEServer(
            List<string> Urls,
            string? Username = null,
            string? Credential = null
        );

        private record LoginRequest(string Username, string Password);
        private record LoginResponse(string Token, User User);
        private record CreateSessionRequest(string RoomName);
        private record JoinSessionRequest(string SessionId);

        // ====================================================================
        // Exceptions
        // ====================================================================

        public class SmaRTCException : Exception
        {
            public SmaRTCException(string message) : base(message) { }
            public SmaRTCException(string message, Exception inner) : base(message, inner) { }
        }

        public class AuthenticationException : SmaRTCException
        {
            public AuthenticationException() : base("Identifiants incorrects") { }
        }

        public class SessionNotFoundException : SmaRTCException
        {
            public SessionNotFoundException() : base("Cet appel n'existe pas") { }
        }

        public class NetworkException : SmaRTCException
        {
            public NetworkException(string message) : base($"Problème de connexion : {message}") { }
            public NetworkException(string message, Exception inner) : base($"Problème de connexion : {message}", inner) { }
        }

        // ====================================================================
        // Client principal
        // ====================================================================

        private readonly HttpClient _httpClient;
        private readonly Config _config;
        private string? _token;
        private string? _currentUsername;
        private string? _currentSessionId;

        public bool IsLoggedIn => !string.IsNullOrEmpty(_token);
        public string? CurrentUsername => _currentUsername;
        public string? CurrentSessionId => _currentSessionId;

        public SmaRTCSimple(Config? config = null)
        {
            _config = config ?? new Config();
            _httpClient = new HttpClient
            {
                BaseAddress = new Uri(_config.ApiBaseUrl),
                Timeout = TimeSpan.FromSeconds(_config.TimeoutSeconds)
            };
        }

        // ====================================================================
        // Méthodes privées
        // ====================================================================

        private async Task<T> RequestAsync<T>(
            HttpMethod method,
            string path,
            object? body = null,
            bool requireAuth = false
        )
        {
            try
            {
                var request = new HttpRequestMessage(method, path);

                if (body != null)
                {
                    request.Content = JsonContent.Create(body);
                }

                if (requireAuth && !string.IsNullOrEmpty(_token))
                {
                    request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _token);
                }

                if (_config.EnableLogs)
                {
                    Console.WriteLine($"[SmaRTC] {method} {_config.ApiBaseUrl}{path}");
                }

                var response = await _httpClient.SendAsync(request);

                if (!response.IsSuccessStatusCode)
                {
                    var errorBody = await response.Content.ReadAsStringAsync();
                    throw response.StatusCode switch
                    {
                        System.Net.HttpStatusCode.Unauthorized => new AuthenticationException(),
                        System.Net.HttpStatusCode.NotFound => new SessionNotFoundException(),
                        _ => new SmaRTCException($"Erreur HTTP {(int)response.StatusCode} : {errorBody}")
                    };
                }

                return await response.Content.ReadFromJsonAsync<T>()
                    ?? throw new SmaRTCException("Réponse vide du serveur");
            }
            catch (HttpRequestException ex)
            {
                throw new NetworkException(ex.Message, ex);
            }
            catch (TaskCanceledException ex)
            {
                throw new NetworkException("Timeout", ex);
            }
        }

        // ====================================================================
        // Méthodes publiques
        // ====================================================================

        /// <summary>
        /// Authentifie l'utilisateur
        /// </summary>
        public async Task LoginAsync(string username, string password)
        {
            var request = new LoginRequest(username, password);
            var response = await RequestAsync<LoginResponse>(HttpMethod.Post, "/api/auth/login", request);

            _token = response.Token;
            _currentUsername = response.User.Username;

            if (_config.EnableLogs)
            {
                Console.WriteLine($"[SmaRTC] Connecté en tant que {_currentUsername}");
            }
        }

        /// <summary>
        /// Crée un nouveau compte utilisateur
        /// </summary>
        public async Task<User> RegisterAsync(string username, string password)
        {
            var request = new LoginRequest(username, password);
            return await RequestAsync<User>(HttpMethod.Post, "/api/auth/register", request);
        }

        /// <summary>
        /// Crée un nouvel appel
        /// </summary>
        public async Task<Session> StartCallAsync(string roomName)
        {
            var request = new CreateSessionRequest(roomName);
            var session = await RequestAsync<Session>(HttpMethod.Post, "/api/session", request, true);

            _currentSessionId = session.SessionId;

            if (_config.EnableLogs)
            {
                Console.WriteLine($"[SmaRTC] Appel créé : {session.SessionId}");
            }

            return session;
        }

        /// <summary>
        /// Rejoint un appel existant
        /// </summary>
        public async Task<Session> JoinCallAsync(string sessionId)
        {
            var request = new JoinSessionRequest(sessionId);
            var session = await RequestAsync<Session>(HttpMethod.Post, "/api/session/join", request, true);

            _currentSessionId = session.SessionId;

            if (_config.EnableLogs)
            {
                Console.WriteLine($"[SmaRTC] Appel rejoint : {session.SessionId}");
            }

            return session;
        }

        /// <summary>
        /// Termine l'appel en cours
        /// </summary>
        public async Task EndCallAsync()
        {
            if (string.IsNullOrEmpty(_currentSessionId))
            {
                throw new SmaRTCException("Aucun appel en cours");
            }

            await RequestAsync<object>(HttpMethod.Delete, $"/api/session/{_currentSessionId}", null, true);

            if (_config.EnableLogs)
            {
                Console.WriteLine($"[SmaRTC] Appel terminé : {_currentSessionId}");
            }

            _currentSessionId = null;
        }

        /// <summary>
        /// Liste tous les appels actifs
        /// </summary>
        public async Task<List<Session>> GetAvailableCallsAsync()
        {
            return await RequestAsync<List<Session>>(HttpMethod.Get, "/api/session", null, true);
        }

        /// <summary>
        /// Récupère la configuration STUN/TURN
        /// </summary>
        public async Task<List<ICEServer>> GetICEServersAsync()
        {
            try
            {
                return await RequestAsync<List<ICEServer>>(HttpMethod.Get, "/api/webrtc/ice", null, true);
            }
            catch
            {
                // Fallback vers Google STUN
                return new List<ICEServer>
                {
                    new ICEServer(new List<string> { "stun:stun.l.google.com:19302" })
                };
            }
        }

        /// <summary>
        /// Déconnecte l'utilisateur
        /// </summary>
        public async Task LogoutAsync()
        {
            // Terminer l'appel en cours si existant
            if (!string.IsNullOrEmpty(_currentSessionId))
            {
                try
                {
                    await EndCallAsync();
                }
                catch (Exception ex)
                {
                    if (_config.EnableLogs)
                    {
                        Console.WriteLine($"[SmaRTC] Erreur lors de la fin d'appel : {ex.Message}");
                    }
                }
            }

            _token = null;
            _currentUsername = null;
            _currentSessionId = null;

            if (_config.EnableLogs)
            {
                Console.WriteLine("[SmaRTC] Déconnecté");
            }
        }

        public void Dispose()
        {
            _httpClient?.Dispose();
        }
    }
}

File: SmaRTC-core/sdk/swift/SmaRTCSimple.swift
// 🍎 SmaRTC Swift SDK Simplifié
// Wrapper moderne pour iOS et macOS
import Foundation

/// Configuration du client SmaRTC
public struct SmaRTCConfig {
    public var apiBaseUrl: String
    public var signalServerUrl: String
    public var timeout: TimeInterval
    public var enableLogs: Bool
    
    public init(
        apiBaseUrl: String = "http://localhost:8080",
        signalServerUrl: String = "http://localhost:5001",
        timeout: TimeInterval = 10,
        enableLogs: Bool = false
    ) {
        self.apiBaseUrl = apiBaseUrl
        self.signalServerUrl = signalServerUrl
        self.timeout = timeout
        self.enableLogs = enableLogs
    }
}

// MARK: - Types de données

public struct User: Codable {
    public let id: String
    public let username: String
}

public struct Session: Codable {
    public let sessionId: String
    public let roomName: String
    public let hostUserId: String
    public let participants: [String]
    public let createdAt: String
    public let isActive: Bool
}

public struct ICEServer: Codable {
    public let urls: [String]
    public let username: String?
    public let credential: String?
}

private struct LoginRequest: Codable {
    let username: String
    let password: String
}

private struct LoginResponse: Codable {
    let token: String
    let user: User
}

private struct CreateSessionRequest: Codable {
    let roomName: String
}

private struct JoinSessionRequest: Codable {
    let sessionId: String
}

// MARK: - Erreurs

public enum SmaRTCError: Error, LocalizedError {
    case authentication
    case sessionNotFound
    case network(String)
    case generic(String)
    
    public var errorDescription: String? {
        switch self {
        case .authentication:
            return "Identifiants incorrects"
        case .sessionNotFound:
            return "Cet appel n'existe pas"
        case .network(let message):
            return "Problème de connexion : \(message)"
        case .generic(let message):
            return message
        }
    }
}

// MARK: - Client SmaRTC

@available(iOS 13.0, macOS 10.15, *)
public class SmaRTCSimple {
    
    private let config: SmaRTCConfig
    private let session: URLSession
    private var token: String?
    private var currentUsername: String?
    private var currentSessionId: String?
    
    public var isLoggedIn: Bool {
        return token != nil
    }
    
    public var username: String? {
        return currentUsername
    }
    
    public var sessionId: String? {
        return currentSessionId
    }
    
    public init(config: SmaRTCConfig = SmaRTCConfig()) {
        self.config = config
        
        let configuration = URLSessionConfiguration.default
        configuration.timeoutIntervalForRequest = config.timeout
        self.session = URLSession(configuration: configuration)
    }
    
    // MARK: - Méthodes privées
    
    private func request<T: Decodable>(
        method: String,
        path: String,
        body: Encodable? = nil,
        requireAuth: Bool = false
    ) async throws -> T {
        guard let url = URL(string: "\(config.apiBaseUrl)\(path)") else {
            throw SmaRTCError.generic("URL invalide")
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        if requireAuth, let token = token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        
        if let body = body {
            let encoder = JSONEncoder()
            request.httpBody = try encoder.encode(body)
        }
        
        if config.enableLogs {
            print("[SmaRTC] \(method) \(url)")
        }
        
        do {
            let (data, response) = try await session.data(for: request)
            
            guard let httpResponse = response as? HTTPURLResponse else {
                throw SmaRTCError.network("Réponse invalide")
            }
            
            if httpResponse.statusCode >= 400 {
                switch httpResponse.statusCode {
                case 401:
                    throw SmaRTCError.authentication
                case 404:
                    throw SmaRTCError.sessionNotFound
                default:
                    let errorMessage = String(data: data, encoding: .utf8) ?? "Erreur inconnue"
                    throw SmaRTCError.generic("Erreur HTTP \(httpResponse.statusCode) : \(errorMessage)")
                }
            }
            
            let decoder = JSONDecoder()
            return try decoder.decode(T.self, from: data)
            
        } catch let error as SmaRTCError {
            throw error
        } catch {
            throw SmaRTCError.network(error.localizedDescription)
        }
    }
    
    // MARK: - Méthodes publiques
    
    /// Authentifie l'utilisateur
    public func login(username: String, password: String) async throws {
        let request = LoginRequest(username: username, password: password)
        let response: LoginResponse = try await self.request(
            method: "POST",
            path: "/api/auth/login",
            body: request
        )
        
        self.token = response.token
        self.currentUsername = response.user.username
        
        if config.enableLogs {
            print("[SmaRTC] Connecté en tant que \(response.user.username)")
        }
    }
    
    /// Crée un nouveau compte utilisateur
    public func register(username: String, password: String) async throws -> User {
        let request = LoginRequest(username: username, password: password)
        return try await self.request(
            method: "POST",
            path: "/api/auth/register",
            body: request
        )
    }
    
    /// Crée un nouvel appel
    public func startCall(roomName: String) async throws -> Session {
        let request = CreateSessionRequest(roomName: roomName)
        let session: Session = try await self.request(
            method: "POST",
            path: "/api/session",
            body: request,
            requireAuth: true
        )
        
        self.currentSessionId = session.sessionId
        
        if config.enableLogs {
            print("[SmaRTC] Appel créé : \(session.sessionId)")
        }
        
        return session
    }
    
    /// Rejoint un appel existant
    public func joinCall(sessionId: String) async throws -> Session {
        let request = JoinSessionRequest(sessionId: sessionId)
        let session: Session = try await self.request(
            method: "POST",
            path: "/api/session/join",
            body: request,
            requireAuth: true
        )
        
        self.currentSessionId = session.sessionId
        
        if config.enableLogs {
            print("[SmaRTC] Appel rejoint : \(session.sessionId)")
        }
        
        return session
    }
    
    /// Termine l'appel en cours
    public func endCall() async throws {
        guard let sessionId = currentSessionId else {
            throw SmaRTCError.generic("Aucun appel en cours")
        }
        
        struct EmptyResponse: Codable {}
        let _: EmptyResponse = try await request(
            method: "DELETE",
            path: "/api/session/\(sessionId)",
            requireAuth: true
        )
        
        if config.enableLogs {
            print("[SmaRTC] Appel terminé : \(sessionId)")
        }
        
        self.currentSessionId = nil
    }
    
    /// Liste tous les appels actifs
    public func getAvailableCalls() async throws -> [Session] {
        return try await request(
            method: "GET",
            path: "/api/session",
            requireAuth: true
        )
    }
    
    /// Récupère la configuration STUN/TURN
    public func getICEServers() async throws -> [ICEServer] {
        do {
            return try await request(
                method: "GET",
                path: "/api/webrtc/ice",
                requireAuth: true
            )
        } catch {
            // Fallback vers Google STUN
            return [
                ICEServer(
                    urls: ["stun:stun.l.google.com:19302"],
                    username: nil,
                    credential: nil
                )
            ]
        }
    }
    
    /// Déconnecte l'utilisateur
    public func logout() async {
        // Terminer l'appel en cours si existant
        if currentSessionId != nil {
            do {
                try await endCall()
            } catch {
                if config.enableLogs {
                    print("[SmaRTC] Erreur lors de la fin d'appel : \(error)")
                }
            }
        }
        
        self.token = nil
        self.currentUsername = nil
        self.currentSessionId = nil
        
        if config.enableLogs {
            print("[SmaRTC] Déconnecté")
        }
    }
}
